File: web/src/lib/protocol.ts
export type UserInfo = {
  username: string
  publicKey: string
  online: boolean
}

export type CallSignal = {
  sdp?: string
  type?: RTCSdpType
  candidate?: string
  sdpMid?: string | null
  sdpMLineIndex?: number | null
}

export type Packet = {
  type:
    | 'message'
    | 'voice'
    | 'setting'
    | 'users'
    | 'error'
    | 'call-offer'
    | 'call-answer'
    | 'call-ice'
    | 'call-hangup'
  from?: string
  to?: string
  publicKey?: string
  message?: string
  selfCopy?: string
  voice?: 'on' | 'off'
  users?: UserInfo[]
  error?: string
  signal?: CallSignal
}

export type StoredMessage = {
  id: number
  peer: string
  from: string
  ciphertext: string
  senderPublicKey: string
  createdAt: number
}

export type ICEServer = {
  urls: string | string[]
  username?: string
  credential?: string
}

export function apiURL(path: string, host?: string): string {
  if (host?.trim()) {
    const h = host.replace(/^https?:\/\//, '').replace(/\/$/, '')
    const proto = window.location.protocol === 'https:' ? 'https:' : 'http:'
    return `${proto}//${h}${path}`
  }
  return path
}

export function wsURL(host?: string): string {
  if (host?.trim()) {
    const h = host.replace(/^https?:\/\//, '').replace(/\/$/, '')
    const proto = window.location.protocol === 'https:' ? 'wss:' : 'ws:'
    return `${proto}//${h}/ws`
  }
  const proto = window.location.protocol === 'https:' ? 'wss:' : 'ws:'
  return `${proto}//${window.location.host}/ws`
}

export function isCallPacket(type: Packet['type']): boolean {
  return (
    type === 'call-offer' ||
    type === 'call-answer' ||
    type === 'call-ice' ||
    type === 'call-hangup'
  )
}

File: internal/protocol/protocol.go
package protocol

import "encoding/json"

// Handshake authenticates a WebSocket and registers the session public key.
type Handshake struct {
	Token     string `json:"token"`
	PublicKey string `json:"publicKey"`
}

// UserInfo is a directory entry for 1:1 chat.
type UserInfo struct {
	Username  string `json:"username"`
	PublicKey string `json:"publicKey"`
	Online    bool   `json:"online"`
}

// Packet is the common envelope for post-handshake messages.
type Packet struct {
	Type      string          `json:"type"`
	From      string          `json:"from,omitempty"`
	To        string          `json:"to,omitempty"`
	PublicKey string          `json:"publicKey,omitempty"`
	Message   string          `json:"message,omitempty"`
	Voice     string          `json:"voice,omitempty"`
	Users     []UserInfo      `json:"users,omitempty"`
	Error     string          `json:"error,omitempty"`
	Signal    json.RawMessage `json:"signal,omitempty"` // WebRTC SDP / ICE payload
}

const (
	TypeMessage    = "message"
	TypeVoice      = "voice"
	TypeSetting    = "setting"
	TypeUsers      = "users"
	TypeError      = "error"
	TypeCallOffer  = "call-offer"
	TypeCallAnswer = "call-answer"
	TypeCallIce    = "call-ice"
	TypeCallHangup = "call-hangup"
)

type Transport interface {
	ReadJSON(v any) error
	WriteJSON(v any) error
	Close() error
	RemoteAddr() string
}

func Marshal(v any) ([]byte, error) {
	return json.Marshal(v)
}

func Unmarshal(data []byte, v any) error {
	return json.Unmarshal(data, v)
}

func IsCallType(t string) bool {
	switch t {
	case TypeCallOffer, TypeCallAnswer, TypeCallIce, TypeCallHangup:
		return true
	default:
		return false
	}
}
